### ISSUE TITLE
GM3.1: GM_openInTab returns null rather than a reference to the created window object

### ISSUE THREAD

**Issue by uar-tim-chamness** [greasemonkey/greasemonkey#2134]
Per http://wiki.greasespot.net/GM_openInTab:
"As of Greasemonkey 0.8.2: returns the window object created."

Starting in version 3.0, GM_openInTab appears to return null rather than the window object created.

Test script:

```
// ==UserScript==
// @name        Test_GM_openInTab
// @namespace   test
// @description Test_GM_openInTab
// @include     https://github.com/greasemonkey/greasemonkey/issues/new
// @version     1
// @grant       GM_openInTab
// ==/UserScript==
var handle = GM_openInTab("https://github.com/");
console.log(handle);
```

Greasemonkey 2.3:
GM_openInTab opens a new tab and returns a Window object referencing the new tab.
![gm2 3](https://cloud.githubusercontent.com/assets/3631298/6952656/ccfc311e-d891-11e4-90e8-2169b4c4629d.png)

Greasemonkey 3.0 & 3.1:
GM_openInTab opens a new tab and returns null.
![gm3 0](https://cloud.githubusercontent.com/assets/3631298/6952658/d073eb02-d891-11e4-91a0-991226c05e20.png)

**Comment by janekptacijarabaci** [comment]
Yes, knows about it.
https://github.com/greasemonkey/greasemonkey/issues/2004#issuecomment-57064797
https://github.com/greasemonkey/greasemonkey/blob/master/content/browser.js#L76

**Comment by janekptacijarabaci** [comment]
Also, this is still true:
> A bunch of "unsafe CPOW usage" warnings in windowId.js lines 8, 17, 23, 28
> and browser.js lines 263 and 267 when opening the monkey menu on any site other than about:blank.
https://github.com/greasemonkey/greasemonkey/issues/2004#issuecomment-69473107

For details see:
https://mikeconley.ca/blog/2015/02/17/on-unsafe-cpow-usage-in-firefox-desktop-and-why-is-my-nightly-so-sluggish-with-e10s-enabled/

**Comment by arantius** [comment]
Post E10S I have no idea how to deal with this.  The parent opens the tab.  The child doesn't know which one it was.

**Comment by Ventero** [comment]
In principle it should be possible to solve this, since `window.open` somehow still manages to return a reference to the window it opened, even with e10s enabled and `dom.ipc.processCount` greater than one.

I *think* the reason this works is because `window.open` actually creates the window that's being opened in the same process as the parent window, though I'm really not sure about this. The code in FF that deals with this is extremely convoluted and hard to follow, but thankfully [this blog post](http://mikeconley.ca/blog/2014/04/30/electrolysis-code-spelunking-how-links-open-new-windows-in-firefox/) provides a good introduction.

Maybe it's possible to do something similar with `GM_openInTab`, which would involve somehow getting a reference to an `nsIWindowProvider` in the frame script and using that to create a new tab/window. How one would go about doing that, I have no idea ...

**Comment by arantius** [comment]
At the very back of my mind, for quite some time, I've thought we should switch back to intentionally returning null.  Was the (non-null) value we returned, for the window where we did, ever actually useful?  I don't know what you could hope to do with it besides call `.close()`, but that was never possible (as far as content was concerned, that window was cross-origin and not available).  Right?

I'd want to know there's actually value there, before we got to something that looks like that much effort to restore this "feature" (?).

**Comment by Ventero** [comment]
Calling `w.close()` is also possible if the window it's being called from is equal to `w.opener`, but I don't think we ever took care of having `.opener` set properly. So the only actually useful thing I can think of is cross-window (and cross-domain) communication using `.postMessage()`.

@uar-tim-chamness, what exactly do you need the window object for?

Personally, I'd be fine with killing off the return value. It would definitely keep things much simpler, and I can't think of a legitimate use-case that could not be achieved by including your script on whatever URL is opened and then running code directly in the window, instead of trying to manipulate it externally through the window object reference.

**Comment by uar-tim-chamness** [comment]
Thanks for all the thought you've put into this problem so far.

The only thing we use the window handle for is calling ```.close()``` on it.  In normal circumstances I believe this wouldn't work, as mentioned above, however by setting dom.allow_scripts_to_close_windows it becomes possible.

The particular script where this issue came up is one used internally by the company I work for to automate some tasks (basically a complicated macro).  The close call is used to clean up some extra windows that need to be opened as part of the task.

However, it sounds like I'm probably an edge case here and given the difficulty of adding this functionality back it may be best to just resolve this issue by updating the documentation to indicate that null is returned as of v3.  The idea of running a separate script on the opened URL for the purposes of closing it may work for me too.

**Comment by Ventero** [comment]
Mh, after playing around with this for a bit, it's actually fairly easy to open a window in the same process as the window the script runs in. Something as simple as `return Services.ww.openWindow(aFrame.content, uri.spec, null, null, null);` in GM_openInTab is already enough to restore the return value. This also incidentally takes care of correctly setting `win.opener`.

But of course, there is one downside to this (we're talking about e10s here after all ...), namely the fact that we lose the ability to specify whether the tab should be loaded in the background. Under the hood, `openWindow` eventually arrives at [_openURIInNewTab](https://dxr.allizom.org/mozilla-central/rev/1d4f44ee51668865945db95c5fbd898e71955a33/browser/base/content/browser.js#4706), which only looks at `browser.tabs.loadDivertedInBackground` to determine whether the tab should be loaded in the background or not - no option to specify anything else from the caller.

Maybe there is a way to immediately switch back to the original tab (that does not involve sending a sync message to the chrome process) or some other way to effectively fake it, but that just means we're moving the complication from getting the return value to ensuring the tab loads in the background.

**Comment by janekptacijarabaci** [comment]
See also #1685

**Comment by TimidScript** [comment]
> Was the (non-null) value we returned, for the window where we did, ever actually useful? I don't know what you could hope to do with it besides call .close(), 

Yep. And that's what I need to do :D. I do hope you can fix this.
